lib/nghttp2_buf.c

```diff
@@ -223,13 +223,60 @@ int nghttp2_bufs_wrap_init(nghttp2_bufs *bufs, uint8_t *begin, size_t len,
   return 0;
 }
 
+int nghttp2_bufs_wrap_init2(nghttp2_bufs *bufs, uint8_t **bufs_in, size_t in_len,
+                          size_t buf_len, nghttp2_mem *mem) {
+  size_t i = 0;
+  nghttp2_buf_chain *cur_chain;
+  nghttp2_buf_chain *pre_chain = NULL;
+  nghttp2_buf_chain *head_chain = NULL;
+
+  for (i = 0; i < in_len; ++i) {
+    uint8_t *begin = bufs_in[i];
+
+    cur_chain = nghttp2_mem_malloc(mem, sizeof(nghttp2_buf_chain));
+    if (cur_chain == NULL) {
+      while (head_chain) {
+        nghttp2_mem_free(mem, head_chain);
+        head_chain = head_chain->next;
+      }
+      return NGHTTP2_ERR_NOMEM;
+    }
+
+    cur_chain->next = NULL;
+    nghttp2_buf_wrap_init(&cur_chain->buf, begin, buf_len);
+
+    if (pre_chain) {
+      pre_chain->next = cur_chain;
+    } else {
+      head_chain = cur_chain;
+    }
+
+    pre_chain = cur_chain;
+  }
+
+  bufs->mem = mem;
+  bufs->offset = 0;
+
+  bufs->head = head_chain;
+  bufs->cur = bufs->head;
+
+  bufs->chunk_length = buf_len;
+  bufs->chunk_used = in_len;
+  bufs->max_chunk = in_len;
+  bufs->chunk_keep = in_len;
+
+  return 0;
+}
+
 void nghttp2_bufs_wrap_free(nghttp2_bufs *bufs) {
   if (bufs == NULL) {
     return;
   }
 
-  nghttp2_mem_free(bufs->mem, bufs->head);
-  bufs->head = NULL;
+  while (bufs->head) {
+    nghttp2_mem_free(bufs->mem, bufs->head);
+    bufs->head = bufs->head->next;
```

Comment:
bufs->head was already freed.  Accessing bufs->head->next is heap-use-after-free.